Migrate controls to streams

PhilipAxelrod opened the pull request.

PhilipAxelrod committed:
wip

PhilipAxelrod committed:
wip (potentially buggy additions

PhilipAxelrod committed:
changes including removing tolerance from trapazoidal

PhilipAxelrod committed:
wip add minus and some pid tests

shadaj commented on core/shared/src/test/scala/com/lynbrookrobotics/potassium/streams/StreamTest.scala:
```diff
@@ -299,4 +299,22 @@ class StreamTest extends FunSuite {
+    // subtractand is what to subtract from minuend
+    val (subtractand, pubSubtractand) = Stream.manual[Length]
+
+    val difference = minuend.minus(subtractand)
```
Instead of this, what if we just have a method `-` so that you could do `minuend - subtractand`?

PhilipAxelrod replied:
I'll add a `-` method, but I'd discourage using it because it gives the idea that subtractand and minuend are single values that you can treat like integers

shadaj replied:
That's a good point, now that I think of it we probably just shouldn't even add the `-` method.

PhilipAxelrod committed:
Fix bugs and write unit tests for derivative and integral control

Integral control integrated the current value not the error, and
derivative differentiated the current value not the error.

PhilipAxelrod committed:
finish porting over most tets except frc

PhilipAxelrod committed:
Fix some bugs in unicycle and position tracking tests, integration, and derivation

shadaj committed:
Implement having a root time stream used by children streams

shadaj committed:
Make all unicycle drive control tests pass

PhilipAxelrod committed:
refactor position tracking

shadaj committed:
Make all unicycle drive tests pass

shadaj committed:
Make pure pursuit tests pass

shadaj committed:
Make all commons module unit tests pass

shadaj committed:
Use ConcurrentLinkedQueue to track listeners of a Stream

shadaj committed:
Use Vector to track listeners of a Stream

shadaj committed:
Attempt to force tests to be run sequentially across subprojects

shadaj committed:
Attempt to make Travis tests pass

shadaj committed:
Force Travis to use an environment with more resources

PhilipAxelrod commented on frc/src/main/scala/com/lynbrookrobotics/potassium/frc/TalonEncoder.scala:
```diff
@@ -1,18 +1,25 @@
+                   talonUpdateRate: Time,
+                   conversionFactor: Ratio[Angle, Dimensionless]) {
+  // TODO: Not granteed to be synchronous with other data
+  val angle: Stream[Angle] = Stream.periodic(talonUpdateRate) (
```
Remove Streams, just include Squants returning methods

PhilipAxelrod commented on lighting/src/main/scala/com/lynbrookrobotics/potassium/lighting/LightingComponent.scala:
```diff
@@ -1,14 +1,16 @@
   var debug = false
 
-  override def defaultController: PeriodicSignal[Int] = Signal[Int](0).toPeriodic
+  // TODO: Please review: seems suspicious
```
Ok!

PhilipAxelrod commented on commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/UnicycleCoreTasks.scala:
```diff
@@ -183,42 +186,48 @@ trait UnicycleCoreTasks {
-        hardware.forwardPosition.map(absoluteDistance - _)
+
+        // TODO: requires review
+        hardware.forwardPosition.relativize((initial, current) => {
```
Okay!

PhilipAxelrod commented on commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/UnicycleCoreTasks.scala:
```diff
@@ -281,14 +292,14 @@ trait UnicycleCoreTasks {
-    val positionSlide: PeriodicSignal[Queue[(Angle, Time)]] = hardware.turnPosition.toPeriodic.zipWithTime.sliding(
-      20, (hardware.turnPosition.get, Milliseconds(System.currentTimeMillis()))
-    )
+    // TODO: requires review
```
Never proven to work, don't worry about it

PhilipAxelrod commented on commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllerTests.scala:
```diff
@@ -54,79 +55,97 @@ class PurePursuitControllerTests extends FunSuite {
 
-    val position = Signal.constant(origin).toPeriodic
+    // TODO: Why is this not inferred?
+    val path = hardware.turnPosition.mapToConstant[(Segment, Option[Segment])](
```
explicit types at variable

PhilipAxelrod commented on commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllerTests.scala:
```diff
@@ -54,79 +55,97 @@ class PurePursuitControllerTests extends FunSuite {
-    val path = Signal.constant((Segment(origin, target), None))
+    val target = Point(Feet(0), Feet(1))
+
+    // TODO: types not inferred correctly
```
Okay

PhilipAxelrod commented on commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllerTests.scala:
```diff
@@ -161,44 +180,55 @@ class PurePursuitControllerTests extends FunSuite {
+    val target = Point(Feet(0), Feet(1))
+    val position = hardware.turnPosition.mapToConstant(Point(Feet(1), Feet(1)))
+
+    // TODO: types not inferred correctly
```
Okay

PhilipAxelrod commented on commons/src/test/scala/com/lynbrookrobotics/potassium/commons/drivetrain/UnicycleDriveControlTest.scala:
```diff
@@ -99,13 +121,22 @@ class UnicycleDriveControlTest extends FunSuite {
+      })
+
+      // TODO: this does not trigger an update in function defined in above foreach
+      trigggerClock.apply(period)
```
Fine as it is now

PhilipAxelrod commented on control/src/main/scala/com/lynbrookrobotics/potassium/control/PIDF.scala:
```diff
@@ -26,40 +28,66 @@ case class PIDFConfig[S <: Quantity[S],
+    throw new NotImplementedException()
+  }
+
+  // TODO: Use zipAsync instead of zip as used in minus
```
Use syncTo when  separate target and current periodicity. Add scaladocs to explain

PhilipAxelrod commented on core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/AsyncZippedStream.scala:
```diff
@@ -1,8 +1,13 @@
   override val expectedPeriodicity: ExpectedPeriodicity = parentA.expectedPeriodicity
 
+  // TODO: Review please
+  override val originTimeStream = parentA.originTimeStream
```
remove, itsk

PhilipAxelrod commented on core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/EagerZippedStream.scala:
```diff
@@ -3,6 +3,9 @@ package com.lynbrookrobotics.potassium.streams
 class EagerZippedStream[A, B](parentA: Stream[A], parentB: Stream[B]) extends Stream[(A, B)] {
   override val expectedPeriodicity: ExpectedPeriodicity = NonPeriodic
 
+  // TODO: Review please. Use Option maybe?
```
Should not be null

PhilipAxelrod commented on core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala:
```diff
@@ -62,16 +66,26 @@ abstract class Stream[T] { self =>
     }
   }
 
+  // TODO: requires review
```
RIP currentValue

PhilipAxelrod commented on core/shared/src/main/scala/com/lynbrookrobotics/potassium/streams/Stream.scala:
```diff
@@ -284,6 +306,33 @@ abstract class Stream[T] { self =>
+
+    // scalastyle:off
+    previousValues.zipWithDt.scanLeft(null.asInstanceOf[I]){case (acc, (current3Values, dt)) =>
+      // TODO: review please
```
Probably ok

PhilipAxelrod committed:
Uncomment mocking test clock

PhilipAxelrod commented on model/src/main/scala/com/lynbrookrobotics/potassium/model/simulations/SimulatedTwoSidedHardware.scala:
```diff
@@ -15,10 +16,17 @@ import scala.collection.mutable
+                           angular: AngularVelocity)
+
+// TODO: Already implemented in TwoSidedDrive
+case class TwoSidedDriveForce(left: Force, right: Force)
```
Might be intellij problem

PhilipAxelrod commented on model/src/main/scala/com/lynbrookrobotics/potassium/model/simulations/SimulatedTwoSidedHardware.scala:
```diff
@@ -28,41 +36,27 @@ class SimulatedTwoSidedHardware(constantFriction: Force,
-    // force periodic signals to update
-    DrivetrainPositions.currentValue(dt)
-    position.currentValue(dt)
+    // TODO: gives warning about pure expressions
```
Not true anymore

PhilipAxelrod commented on model/src/main/scala/com/lynbrookrobotics/potassium/model/simulations/SimulatedTwoSidedHardware.scala:
```diff
@@ -79,74 +73,73 @@ class SimulatedTwoSidedHardware(constantFriction: Force,
+    TwoSidedDriveForce(o._1, o._2))
+
+  // TODO:
+  // temporary to make integral and derivative compile because implicit clock
```
Just remove

PhilipAxelrod commented on model/src/main/scala/com/lynbrookrobotics/potassium/model/simulations/SimulatedTwoSidedHardware.scala:
```diff
@@ -28,41 +36,27 @@ class SimulatedTwoSidedHardware(constantFriction: Force,
   val history = new mutable.ArrayBuffer[MomentInHistory]
```
Move to Stream of MomentInHistory

shadaj committed:
Use correct package for ClockMockingTest

Commit:
Make encodes use basic interface instead of streams greater flexibility

PhilipAxelrod committed:
remove print statements

PhilipAxelrod committed:
Disaccociate digital gryo from Signals. Fix 1 of several open  bugs causing simulated drivetrain not to update

shadaj committed:
Fix critical bug with zipped streams getting cleared out by GC

shadaj committed:
Enable PIDFTest and remove debug messages in Stream

shadaj committed:
Implement more optimized `drop` method on Stream

Commit:
Migrate implicits to streams

PhilipAxelrod committed:
Make interface tests pass and add benchmark for publishing to zipped streams

PhilipAxelrod committed:
Remove extra white spaces

PhilipAxelrod committed:
remove unnecessary TODO's

PhilipAxelrod committed:
Add task to drive at angle. Refactor position tracking tests

PhilipAxelrod committed:
Remove extra TODO's

shadaj commented:
After discussion with @Paxelord we have decided that we will move towards merging this, with additional migrations in separate PRs.

shadaj reviewed (approved):
LGTM!
